### PR TITLE
clp-s: Add support for chunking output into different files during timestamp-ordered decompression

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR adds support for chunking the output of timestamp-ordered decompression into several files, where each file has at most the number of records specified in the command line argument. The argument `--ordered-chunk-split-threshold <value>` can be used in conjunction with the `--ordered` argument during decompression to trigger this feature.

# Validation performed
- Tested edge case where every records ends up in same chunk
- Tested edge case where num_records % chunk_size == 0
- Tested case where num_records % chunk_size > 0

### CODE REVIEW

File: components/core/src/clp_s/JsonConstructor.cpp [hunk 1]
```diff
@@ -72,15 +74,60 @@ void JsonConstructor::construct_in_order(FileWriter& writer) {
...
+
+    epochtime_t first_timestamp{0}, last_timestamp{0};
+    size_t num_records_marshalled{0};
+    auto src_path = std::filesystem::path(m_output_dir) / m_archive_id;
+    FileWriter writer;
+    writer.open(src_path, FileWriter::OpenMode::CreateForWriting);
+
+    auto finish_chunk = [&](bool open_new_writer) {
+        writer.close();
+        std::string new_file_name = std::string(src_path) + "_" + std::to_string(first_timestamp)
```
Comment (wraymo) [inline components/core/src/clp_s/JsonConstructor.cpp:86]:
```suggestion
        std::string new_file_name = src_path.string() + "_" + std::to_string(first_timestamp)
```

File: components/core/src/clp_s/JsonConstructor.cpp [hunk 2]
```diff
@@ -72,15 +74,60 @@ void JsonConstructor::construct_in_order(FileWriter& writer) {
...
     // a given table
     tables.clear();
+
+    epochtime_t first_timestamp{0}, last_timestamp{0};
+    size_t num_records_marshalled{0};
+    auto src_path = std::filesystem::path(m_output_dir) / m_archive_id;
+    FileWriter writer;
+    writer.open(src_path, FileWriter::OpenMode::CreateForWriting);
+
+    auto finish_chunk = [&](bool open_new_writer) {
```
Comment (wraymo) [inline components/core/src/clp_s/JsonConstructor.cpp:84]:
```suggestion
    auto finalize_chunk = [&](bool open_new_writer) {
```

File: components/core/src/clp_s/CommandLineArguments.hpp [hunk 3]
```diff
@@ -170,6 +172,7 @@ class CommandLineArguments {
     size_t m_max_document_size{512ULL * 1024 * 1024};  // 512 MB
     bool m_structurize_arrays{false};
     bool m_ordered_decompression{false};
+    size_t m_ordered_chunk_split_threshold{0};
```
Comment (wraymo) [inline components/core/src/clp_s/CommandLineArguments.hpp:175]:
```suggestion
    size_t m_ordered_chunk_size{0};
```

File: components/core/src/clp_s/CommandLineArguments.cpp [hunk 4]
```diff
@@ -287,11 +287,20 @@ CommandLineArguments::parse_arguments(int argc, char const** argv) {
...
             decompression_options.add_options()(
                     "ordered",
                     po::bool_switch(&m_ordered_decompression),
                     "Enable decompression in ascending timestamp order for this archive"
+            )(
+                    "ordered-chunk-split-threshold",
+                    po::value<size_t>(&m_ordered_chunk_split_threshold)
+                            ->default_value(m_ordered_chunk_split_threshold),
+                    "Number of records to include in each output chunk when decompressing records "
+                    "in order"
```
Comment (wraymo) [inline components/core/src/clp_s/CommandLineArguments.cpp:300]:
```suggestion
                    "in timestamp ascending order"
```

File: components/core/src/clp_s/CommandLineArguments.cpp [hunk 5]
```diff
@@ -287,11 +287,20 @@ CommandLineArguments::parse_arguments(int argc, char const** argv) {
...
+            // clang-format off
             decompression_options.add_options()(
                     "ordered",
                     po::bool_switch(&m_ordered_decompression),
                     "Enable decompression in ascending timestamp order for this archive"
+            )(
+                    "ordered-chunk-split-threshold",
+                    po::value<size_t>(&m_ordered_chunk_split_threshold)
+                            ->default_value(m_ordered_chunk_split_threshold),
+                    "Number of records to include in each output chunk when decompressing records "
```
Comment (wraymo) [inline components/core/src/clp_s/CommandLineArguments.cpp:299]:
```suggestion
                    "Number of records to include in each output file when decompressing records "
```

**Review by wraymo — COMMENTED** [review]
Great work! Most of the comments are about the style changes.

File: components/core/src/clp_s/JsonConstructor.cpp [hunk 6]
```diff
@@ -72,15 +74,60 @@ void JsonConstructor::construct_in_order(FileWriter& writer) {
...
+        std::error_code ec;
+        std::filesystem::rename(src_path, new_file_path, ec);
+        if (ec) {
+            throw OperationFailed(ErrorCodeFailure, __FILE__, __LINE__, ec.message());
+        }
+
+        if (open_new_writer) {
+            writer.open(src_path, FileWriter::OpenMode::CreateForWriting);
+        }
+    };
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp_s/JsonConstructor.cpp:98]: I think it's easier to read this way, but if you prefer private method I can change it.

File: components/core/src/clp_s/CommandLineArguments.cpp (hunk 4 above)
Comment (gibber9809) [reply components/core/src/clp_s/CommandLineArguments.cpp]: I replaced with "in ascending timestamp order" instead.

File: components/core/src/clp_s/CommandLineArguments.cpp [hunk 7]
```diff
@@ -287,11 +287,19 @@ CommandLineArguments::parse_arguments(int argc, char const** argv) {
             extraction_options.add(input_options);
 
             po::options_description decompression_options("Decompression Options");
+            // clang-format off
             decompression_options.add_options()(
                     "ordered",
                     po::bool_switch(&m_ordered_decompression),
                     "Enable decompression in ascending timestamp order for this archive"
+            )(
+                    "ordered-chunk-split-threshold",
```
Comment (wraymo) [inline components/core/src/clp_s/CommandLineArguments.cpp:296]: Do you want to update the name of this argument?

**Review by wraymo — APPROVED** [review]
For the PR title, what about "clp-s: Add support for chunking output into different files during timestamp-ordered decompression"?

**Pull request merged**